### PR TITLE
[mlir python] Add nanobind support for standalone dialects.

### PR DESCRIPTION
**Pull request by hawkinsp**
This PR allows out-of-tree dialects to write Python dialect modules using nanobind instead of pybind11.

It may make sense to migrate in-tree dialects and some of the ODS Python infrastructure to nanobind, but that is a topic for a future change.

This PR makes the following changes:
* adds nanobind to the CMake and Bazel build systems. We also add robin_map to the Bazel build, which is a dependency of nanobind.
* adds a PYTHON_BINDING_LIBRARY option to various CMake functions, such as declare_mlir_python_extension, allowing users to select a Python binding library.
* creates a fork of mlir/include/mlir/Bindings/Python/PybindAdaptors.h named NanobindAdaptors.h. This plays the same role, using nanobind instead of pybind11.
* splits CollectDiagnosticsToStringScope out of PybindAdaptors.h and into a new header mlir/include/mlir/Bindings/Python/Diagnostics.h, since it is code that is no way related to pybind11 or for that matter, Python.
* changed the standalone Python extension example to have both pybind11 and nanobind variants.
* changed mlir/python/mlir/dialects/python_test.py to have both pybind11 and nanobind variants.

Notes:
* A slightly unfortunate thing that I needed to do in the CMake integration was to use FindPython in addition to FindPython3, since nanobind's CMake integration expects the Python_ names for variables. Perhaps there's a better way to do this.

### CODE REVIEW

**Comment by stellaraccident** [comment]
Oh man, hit me with this four hours before I'm planning a two week vacation :)

I'll take a look now. Generally supportive. Also supportive of migrating the whole to nanobind given that it is such better performance and ergonomics, but as you say -- a topic for a different day.

File: mlir/cmake/modules/MLIRDetectPythonEnv.cmake
Comment (stellaraccident) [inline mlir/cmake/modules/MLIRDetectPythonEnv.cmake:26]: Agreed with the comment. I've run into this too but have also just done the pragmatic thing. Might be worth filing an issue in nanobind for guidance.

**Review by stellaraccident — COMMENTED** [review]
If history is a guide, this may be challenging to land because of build bot deps needing to be updated. Will see what the bots say about that.

One practical thing: every time we touch MLIRDetectPythonEnv, it causes a lot of pain and suffering downstream, and this is going to be another dep change/expansion that will trigger that. I think the way out in those cases may just be to add a CMake option to have MLIR inherit the python configuration from a super-project vs trying to probe and configure everything itself. Then such projects can set that option and just do it themselves vs fighting with MLIR over how to do the detection.

I'm going to fiddle with this now as it has been on the agenda for a long time.

**Comment by stellaraccident** [comment]
Please add `nanobind>=2.0` to `mlir/python/requirements.txt`. This is how we document that users should get their deps. I can't remember if the bots use it (but probably should).

**Comment by hawkinsp** [comment]
> Please add `nanobind>=2.0` to `mlir/python/requirements.txt`. This is how we document that users should get their deps. I can't remember if the bots use it (but probably should).

Already there.

BTW, is there documentation on how to run all the MLIR Python tests? Apparently CI is running some I didn't find. I looked at https://mlir.llvm.org/docs/Bindings/Python/#testing for guidance but to no avail.

**Comment by stellaraccident** [comment]
Mind having a look at: https://github.com/llvm/llvm-project/pull/117934

If we land that, I'll send out a PSA to users to enable the new option if they have a project that has an opinionated setup of their Python dev packages and give us one less thing to worry about landing this one.

**Comment by stellaraccident** [comment]
I'm going to need to hand you over to @makslevental to get this in next week, but this looks reasonable to me.

I think that we should land #117934 first and possibly send out a note telling people with non trivial super projects to use that option. That would require some minor indent-rebasing with this patch.

**Comment by joker-eph** [comment]
I don't think the bots are using the requirements.txt from the repo still (unfortunately), we'll likely have to manually re-deploy images instead.

**Comment by stellaraccident** [comment]
> I don't think the bots are using the requirements.txt from the repo still (unfortunately), we'll likely have to manually re-deploy images instead.

I thought that as well, but it appears to be: https://github.com/llvm/llvm-project/blob/main/.ci/monolithic-linux.sh#L48

Is there somewhere else to look?

**Comment by joker-eph** [comment]
This is the pre-merge CI, which runs on the PR right now.

I was referring to the various buildbot which runs post-merge, the configs are defined in https://github.com/llvm/llvm-zorg/

**Comment by stellaraccident** [comment]
> This is the pre-merge CI, which runs on the PR right now.
> 
> I was referring to the various buildbot which runs post-merge, the configs are defined in https://github.com/llvm/llvm-zorg/

Ah. I knew I was remembering something. Well, I'm about to be offline/vacation so will leave that to next week.

**Comment by makslevental** [comment]
FWIW I think we could just wholesale switch to `nanobind`. Prior, I wasn't sure what the feature disparities were but I've been writing a new set of bindings this week and it looks to me like it's all there and basically spelled exactly the same.

**Comment by stellaraccident** [comment]
> FWIW I think we could just wholesale switch to `nanobind`. Prior, I wasn't sure what the feature disparities were but I've been writing a new set of bindings this week and it looks to me like it's all there and basically spelled exactly the same.

There are differences that I've bridged by writing various helpers, mostly around array and buffer protocols. Then the way that constructors are modeled is wholly better but different. I found that when retrofitting, there are usually a few places that leave you scratching your head on how to do the same thing. But when writing fresh, the nanobind way is more powerful. 

I think it's kind of nice to do what Peter is doing here and getting it in universe first. Removes some activation energy for taking it the rest of the way.

**Comment by stellaraccident** [comment]
FYI: https://discourse.llvm.org/t/psa-python-binding-dependencies-changing/83376

**Comment by jpienaar** [comment]
+1 that this has been on the wishlist for a while to migrate.

**Comment by hawkinsp** [comment]
I think I resolved the remaining test failures. Any further action items for me?

**Comment by makslevental** [comment]
@hawkinsp basically looks good to me; two questions (one load bearing, one curiosity):

1. We currently (in the pybind impl) have `module_local` in lots of places. nanobind doesn't have `module_local` and instead uses `NB_DOMAIN` (which maybe works the [same](https://github.com/wjakob/nanobind/blob/master/include/nanobind/nb_defs.h#L190) [way](https://github.com/wjakob/nanobind/blob/master/src/nb_internals.cpp#L384-L385)???). Do we want to support that? 
2. Do the added bazel functions work in OSS? Possibly the question is ill-posed because scanning through the bazel I don't see a problematic rule that I thought I remembered seeing (`pybind_extension`) but I'm actually asking on behalf of users like @j2kun who have had trouble using python bindings bazel infra for OSS projects.

**Comment by hawkinsp** [comment]
> @hawkinsp basically looks good to me; two questions (one load bearing, one curiosity):
> 
> 1. We currently (in the pybind impl) have `module_local` in lots of places. nanobind doesn't have `module_local` and instead uses `NB_DOMAIN` (which maybe works the [same](https://github.com/wjakob/nanobind/blob/master/include/nanobind/nb_defs.h#L190) [way](https://github.com/wjakob/nanobind/blob/master/src/nb_internals.cpp#L384-L385)???). Do we want to support that?

Well, this is a choice for whoever builds these extensions. In the most recent push I added an `NB_DOMAIN` of `mlir` to the CMake rules in this PR, which perhaps achieves what you want, i.e., a unique namespace for the upstream mlir bindings, when built in their default configuration?

It doesn't really matter at the moment because the only artifacts built with nanobind are the standalone example and the python test dialect, but it will matter in the future if we build, e.g., the ODS artifacts with nanobind.

> 2. Do the added bazel functions work in OSS? Possibly the question is ill-posed because scanning through the bazel I don't see a problematic rule that I thought I remembered seeing (`pybind_extension`) but I'm actually asking on behalf of users like @j2kun who have had trouble using python bindings bazel infra for OSS projects.

I verified:
```
bazel build --config=generic_clang @llvm-project//...
```

works with this PR included in OSS.

`pybind_extension` is frequently misnamed. To build a pybind11 extension, you need basically to build a shared library with: (a) to enable exceptions, if they aren't, (b) to depend on pybind11. Nanobind is exactly the same, just depend on nanobind instead of pybind11.

File: mlir/cmake/modules/AddMLIRPython.cmake
Comment (makslevental) [inline mlir/cmake/modules/AddMLIRPython.cmake:661]: I was imagining this would be an arg (that naturally defaulted to `mlir`) so that downstream users could set this to whatever they wanted (if they wanted).

**Review by makslevental — APPROVED** [review]
Thanks @hawkinsp - looking forward to more `nb::...` :smile:

File: mlir/cmake/modules/AddMLIRPython.cmake
Comment (hawkinsp) [reply mlir/cmake/modules/AddMLIRPython.cmake:661]: Let's do that in a follow-on change?

**Comment by hawkinsp** [comment]
> Thanks @hawkinsp - looking forward to more `nb::...` 😄

I'm actually a pretty good fraction of the way through porting the MLIR Python core to nanobind as well, and I've made enough progress that I can get the JAX CPU test suite to pass. It has mostly been mechanical, with a few small changes to deal with various semantic differences. I'll open another PR for that when it's closer to ready, and I guess there's some sort of RFC process we'll need to follow as well.

File: mlir/cmake/modules/AddMLIRPython.cmake
Comment (makslevental) [reply mlir/cmake/modules/AddMLIRPython.cmake:661]: Sure np

**Comment by stellaraccident** [comment]
That's a pretty material change, so we should make an RFC. The implementation choices have never been exposed as part of the public API, so I don't expect it to be controversial. But a number of folks do build on this stuff and need to know. 

I'm on vacation and trying to be offline for the next two weeks, but if its after that, I can help smooth anything that needs smoothing. We'll probably also want to give a window for downstream testing to catch anything unexpected. 

If its before that, maks can tune in.

Thanks for the labor. Switching will be a big improvement.

**Comment by makslevental** [comment]
> I'll open another PR for that when it's closer to ready, and I guess there's some sort of RFC process we'll need to follow as well.

I don't think you'll need an RFC - it's an upgrade right? Anyway even if there is an RFC, I don't think it will be controversial.

**Comment by hawkinsp** [comment]
@makslevental can you please merge this? Thanks!

**Comment by hawkinsp** [comment]
Hmm. It looks like some of the build bots don't have the extra dep, as folks guessed above. How does one fix that?

**Comment by makslevental** [comment]
Looks like the pip install deps are here https://github.com/llvm/llvm-zorg/blob/fab27e270d1a3754c98175d780bc4793e13fefc5/buildbot/mlir-nvidia/Dockerfile#L53 so that needs a PR but I don't know if that triggers a rebuild of the container? For a moment I'm gonna revert.

**Comment by makslevental** [comment]
created https://github.com/llvm/llvm-zorg/pull/327 for the nvidia changes but the rocm failure isn't so simple (I don't see a Dockerfile for that bot).

**Comment by hawkinsp** [comment]
If the configuration isn't in the zorg repository, I'd assume the only other thing we can do is ping the owner of the bot?

**Comment by makslevental** [comment]
Okay I found the AMD infra [/utils/buildbot/mlir-rocm-mi200](https://github.com/ROCm/rocMLIR/tree/e498eadb154315bc5c72d6cc85f2ffcdd146b4cc/mlir/utils/buildbot/mlir-rocm-mi200) but it needs to be rebuilt I guess.

> I'd assume the only other thing we can do is ping the owner of the bot?

I believe the owner is no longer at AMD but I am so I can ask aroud.

**Comment by nikic** [comment]
Does nanobind have to be a required dependency?

**Comment by joker-eph** [comment]
I think we're [migrating from pybind11 to nanobind](https://discourse.llvm.org/t/nanobind-for-mlir-python-bindings/83511), so using the python bindings will need nanobind (is this what you're asking?).

**Comment by hawkinsp** [comment]
> Does nanobind have to be a required dependency?

Yes, that's correct. If it were just this PR, it would be reasonable for it to be optional. We'd certainly want to *test* it in CI, although it would be fine if a user didn't have it and didn't want it.

However, note https://github.com/llvm/llvm-project/pull/118583, which migrates the Python core code to nanobind, at which point it would certainly be mandatory. Indeed at that point the question is "can pybind11 be optional"?

Is there a way to specify dependencies that should be present for CI tests but are not mandatory for users?

**Comment by stellaraccident** [comment]
> Is there a way to specify dependencies that should be present for CI tests but are not mandatory for users?

Just documentation and tribal knowledge on how to be a downstream in a non trivial way. The MLIR requirements.txt file describes the deps that a user must have to build and run `check-mlir` without error. There is not a corresponding thing that systematically tells an integrator of the ways they can slice it further. I'd vote for better documenting the end state for integrators, when pybind is optional vs now.

**Comment by nikic** [comment]
Okay, if the python bindings are going to use this internally (rather than just make it an option for consumers), then making it required seems unavoidable.

The reason I asked is that nanobind is currently not packaged in Fedora, which probably means that MLIR is the first significant user of this library. (Work to package it is underway now...)

**Comment by kwk** [comment]
> (Work to package it is underway now...)

The package source code is here: https://github.com/kwk/python-nanobind/
And the package review request is here: https://bugzilla.redhat.com/show_bug.cgi?id=2331339

**Comment by stellaraccident** [comment]
The other thing we could do is to just get it with FetchContent vs relying on os packages. We really can't be version locked to what the OS's do for this.

We use FetchContent for this reason downstream and it works best. We can update mlir itself to that I think (need to double check what version of cmake llvm pins to as that is usually what keeps us from having nice things).

In any case, I'm very -1 on pinning use of nanobind to os packages.

**Comment by nikic** [comment]
In order for distros to ship MLIR python bindings, the dependencies must also be provided by the distro. For obvious reasons, it's not possible to just pull in code over the network during the build.

**Comment by stellaraccident** [comment]
If that is a hard requirement, we will probably need to exclude the python packages from distribution in OS packages (they were never scoped for this anyway) or move them out of tree. The python ecosystem is simply moving too fast to gate on this. We cannot stay in sync with python.org development without being able to use bleeding edge deps in the bindings.

We could also vendor them into the monorepo if there was a strong push from the OS packaging side to distribute these demo bindings. I can't overstate how inconsequential this decision should be from a packaging standpoint.

I'll note that there is an RFC for this, which is a better place to have the discussion (I'm not sure many are monitoring this merged PR): 

https://discourse.llvm.org/t/nanobind-for-mlir-python-bindings/83511

**Comment by marbre** [comment]
> If that is a hard requirement, we will probably need to exclude the python packages from distribution in OS packages (they were never scoped for this anyway) or move them out of tree. The python ecosystem is simply moving too fast to gate on this. We cannot stay in sync with python.org development without being able to use bleeding edge deps in the bindings.

From packaging for Linux distributions, I agree that there is a strong need to only consume dependencies that are packaged for the OS. However, as long as one can disable `FetchContent` without needing to patch the `CMakeLists.txt` by a build option it should okay to provide getting nanobind via `FetchContent` as an option. With this it isn't enforced and distributions can build upon what is shipped for the OS. If they don't ship the dep they still can exclude the python package to not pull in a dependency during build time.

**Comment by nikic** [comment]
Right -- the version of nanobind packaged by the distro is only really a concern for the MLIR bindings packaged by the distro. People building MLIR themselves are free to install a newer version, e.g. using pip and requirements.txt.

**Pull request merged**
